sonic_data_client/events_client.go

```diff
@@ -0,0 +1,306 @@
+
+const STATS_FIELD_NAME = "value"
+
+
+type EventClient struct {
+
+    prefix      *gnmipb.Path
+    path        *gnmipb.Path
+
+    q           *queue.PriorityQueue
+    channel     chan struct{}
+
+    wg          *sync.WaitGroup // wait for all sub go routines to finish
+
+    subs_handle unsafe.Pointer
+
+    stopped     int
+
+    // Stats counter
+    counters    map[string]uint64
+
+    last_latencies  [LATENCT_LIST_SIZE]int64;
+    last_latency_index  int
+    last_latency_full   bool
+
+    last_errors int64
+}
+
+func NewEventClient(paths []*gnmipb.Path, prefix *gnmipb.Path, logLevel int) (Client, error) {
+    var evtc EventClient
+    evtc.prefix = prefix
+    for _, path := range paths {
+        // Only one path is expected. Take the last if many
+        evtc.path = path
+    }
+    C.swssSetLogPriority(C.int(logLevel))
+
+    /* Init subscriber with cache use and defined time out */
+    evtc.subs_handle = C.events_init_subscriber_wrap(true, C.int(SUBSCRIBER_TIMEOUT))
+    evtc.stopped = 0
+
+    /* Init list & counters */
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for _, key := range STATS_ABSOLUTE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for i := 0, i < LATENCT_LIST_SIZE; i++ {
+        evtc.last_latencies[i] = 0
+    }
+    evtc.last_latency_index = 0
+    evtc.last_errors = 0
+    evtc.last_latency_full = false;
+
+    log.V(7).Infof("NewEventClient constructed. logLevel=%d", logLevel)
+
+    return &evtc, nil
+}
+
+
+func update_stats(evtc *EventClient) {
+    ns := sdcfg.GetDbDefaultNamespace()
+    rclient := redis.NewClient(&redis.Options{
+        Network:    "tcp",
+        Addr:       sdcfg.GetDbTcpAddr("COUNTERS_DB", ns),
+        Password:   "", // no password set
+        DB:         sdcfg.GetDbId("COUNTERS_DB", ns)
+        DialTimeout:0,
+    })
+
+    var db_counters map[string]uint64
+    var wr_counters *map[string]uint64 = NULL;
+
+    /* Init current values for cumulative keys and clear for absolute */
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        fv, err := rclient.HGetAll(key).Result()
+        db_counters[key] = fv[STATS_FIELD_NAME]
+    }
+    for _, key := range STATS_ABSOLUTE_KEYS {
+        db_counters[key] = 0;
+    }
+
+    for evtc.stopped == 0 {
+        var tmp_counters map[string]uint64
+
+        // compute latency
+        if evtc.last_latency_full {
+            var total uint64 = 0
+            var cnt int = 0
+
+            for _ v := range evtc.last_latencies {
+                if v > 0 {
+                    total += v
+                    cnt++
+                }
+            }
+            evtc.counters[LATENCY] = (int64) (total/cnt/1000/1000)
+        }
+
+        for key, val := range evtc.counters {
+            tmp_counters[key] = val + db_counters[key]
+        }
+        tmp_counters[DROPPED] += evtc.last_errors
+
+        if (wr_counters != nil) && !reflect.DeepEqual(tmp_counters, *wr_counters) {
```

Comment:
Can you add some comments for this block of code? When is wr_counters not nil? Is line 157 supposed to be outside of this block?